fvm/state/accounts_status.go

```diff
@@ -1,51 +1,112 @@
 package state
 
 import (
+	"encoding/binary"
 	"encoding/hex"
 
+	"github.com/onflow/atree"
+
 	"github.com/onflow/flow-go/fvm/errors"
 )
 
-type AccountStatus uint8
+const (
+	flagSize            = 1
+	storageUsedSize     = 8
+	storageIndexSize    = 8
+	publicKeyCountsSize = 8
+
+	accountStatusSize = flagSize +
+		storageUsedSize +
+		storageIndexSize +
+		publicKeyCountsSize
+
+	flagIndex                 = 0
+	storageUsedStartIndex     = flagIndex + flagSize
+	storageIndexStartIndex    = storageUsedStartIndex + storageUsedSize
+	publicKeyCountsStartIndex = storageIndexStartIndex + storageIndexSize
+)
+
+// AccountStatus holds meta information about an account
+//
+// currently modelled as a byte array with on-demand encoding/decoding of sub arrays
+// the first byte captures flags (e.g. frozen)
+// the next 8 bytes (big-endian) captures storage used by an account
+// the next 8 bytes (big-endian) captures the storage index of an account
+// and the last 8 bytes (big-endian) captures the number of public keys stored on this account
+type AccountStatus [accountStatusSize]byte
 
 const (
-	maskExist  byte = 0b0000_0001
 	maskFrozen byte = 0b1000_0000
 )
 
-// NewAccountStatus sets exist flag and return an AccountStatus
-func NewAccountStatus() AccountStatus {
-	return AccountStatus(maskExist)
+// NewAccountStatus returns a new AccountStatus
+func NewAccountStatus(storageIndex atree.StorageIndex) *AccountStatus {
+	var as AccountStatus
+	as.SetStorageIndex(storageIndex)
+	return &as
 }
 
-func (a AccountStatus) ToBytes() []byte {
-	b := make([]byte, 1)
-	b[0] = byte(a)
-	return b
+// ToBytes converts AccountStatus to a byte slice
+//
+// this has been kept this way in case one day
+// we decided to move on to use an struct to represent
+// account status.
+func (a *AccountStatus) ToBytes() []byte {
+	return a[:]
 }
 
-func AccountStatusFromBytes(inp []byte) (AccountStatus, error) {
-	// if len of inp is zero, account does not exist
-	if len(inp) == 0 {
-		return 0, nil
+// AccountStatusFromBytes constructs an AccountStatus from the given byte slice
+func AccountStatusFromBytes(inp []byte) (*AccountStatus, error) {
+	var as AccountStatus
+	if len(inp) != accountStatusSize {
+		return &as, errors.NewValueErrorf(hex.EncodeToString(inp), "invalid account status size")
 	}
-	if len(inp) > 1 {
-		return 0, errors.NewValueErrorf(hex.EncodeToString(inp), "invalid account state")
+	copy(as[:], inp)
+	return &as, nil
```

Comment:
Maybe we can avoid making a copy by using input's underlying array.  And we don't need `as` created earlier.

```suggestion
	return (*AccountStatusArray)(inp), nil
```

Reply:
I was not sure, I was worried we might update the registers directly in this case and if we get rid of deep copy feature of DB we might have some issues? what do you think ?

Reply:
>I was worried we might update the registers directly in this case

I think it's OK if `AccountStatusFromBytes` is only called to create account status by reading:
- payload value from mtrie, or
- cached payload value, previously retrieved from mtrie.

Because [a copy of payload value is returned from mtrie](
https://github.com/onflow/flow-go/blob/01fae8ea3b6658715880aa014866e9b8b61233bb/ledger/complete/mtrie/forest.go#L135), I think we update the copy.

>if we get rid of deep copy feature of DB we might have some issues

Yes but getting rid of deep copy of the payload value can be problematic even without this change.  I think it would be safer to get rid of deep copy of the payload key.

Reply:
Another case I still not sure, is the case we rollback, we drop the delta but the underlying read cache on view is invalid.

Reply:
let's keep this optimization for another PR when we revisit all allocations and possible ways to not copy or deep copy.